support subset embedding of OpenType CFF fonts

matsud224 opened the pull request:
This PR adds a subset embedding support of OpenType CFF fonts to otfm.
This patch includes some encoders for CFF table. To reduce the embedded font size, unnecessary Font DICTs and subroutines are also omitted. 

I tested this patch with lmmono10-regular, [NotoSansCJKjp-Regular](https://github.com/googlei18n/noto-fonts/blob/master/hinted/NotoSansMono-Regular.ttf), [SourceHanSerif-Regular](https://github.com/adobe-fonts/source-han-serif/tree/release/OTF/Japanese), and [OpenType Collection version of SourceHanSerif-Regular](https://github.com/adobe-fonts/source-han-serif/tree/release/OTC).

matsud224 committed:
add encoder for cff table (header - charstrings index)

matsud224 committed:
add encoder for cff table (font dict index - local subr index)

matsud224 committed:
refactor

matsud224 committed:
fix Makefile and _tags

matsud224 committed:
fix integer encoding

matsud224 committed:
add maxp table version 0.5 support

matsud224 committed:
introduce outline_type

matsud224 committed:
fix singleton index encoding

matsud224 committed:
fix calculation of real number length

matsud224 committed:
add fixed-length integer encoding option

matsud224 committed:
fix charstrings offset in topdict

matsud224 committed:
fix encoding of CharStringData

matsud224 committed:
fix lsubr offset

matsud224 committed:
fix offset calculation of local subrs

matsud224 committed:
fix bug of lsubr search

matsud224 committed:
impl usused font dict removal

matsud224 committed:
fix fdselect

matsud224 committed:
fix not to generate fdindex for non-CIDFonts

matsud224 committed:
add unused subrs removal

matsud224 committed:
fix unused to used

matsud224 committed:
fix subroutine number

matsud224 committed:
fix invalid glyph offset bug

matsud224 committed:
fix invalid key

matsud224 committed:
fix bug of not using reduced lsubridx

matsud224 committed:
fix compare function of SubrIndexMap

matsud224 committed:
remove unnecessary code

matsud224 commented:
I confirmed that generated PDFs are correctly rendered on evince, mupdf, google-chrome and xpdf, but pdf.js (Firefox) couldn't render them. (Glyphs of NotoSansCJK and SourceHansSerif are not completely shown.)

I'll inspect this problem...

matsud224 committed:
fix dictionary update

matsud224 committed:
add charset

matsud224 committed:
fix enc_charset_identity

matsud224 commented:
The problem described above was solved by adding Charset into CFF table (commit `
4d808c2` - `09d3506`).

gfngfn commented:
Thank you very much for making such a wonderful improvement! (And I am sorry for the late response mainly due to personal reasons such as house-moving.)

Unfortunately, it seems there are still some problems that slightly prevent the intended operation… I have encountered the following rendered results of a PDF outputted by using `otfm` enhanced by this PR:

<img width="843" alt="スクリーンショット 2019-03-27 17 43 46" src="https://user-images.githubusercontent.com/6876129/55061875-26ff9780-50b8-11e9-8a55-8439bd53159c.png">
<img width="774" alt="スクリーンショット 2019-03-27 17 44 36" src="https://user-images.githubusercontent.com/6876129/55061904-367ee080-50b8-11e9-9662-04138dc90370.png">

Both rendered results come from a single PDF; The first one is produced by `Preview.app`, while the second one is by Adobe Acrobat Reader DC (for macOS). Here is the input for SATySFi to gain the PDF:

```
@require: stdjabook
@require: itemize

let-block ctx +p-hira it =
  let ctx = ctx |> StdJaBook.set-cjk-font (`HiraKakuProN-W2`, 0.88, 0.) in
  line-break true true ctx
    (inline-skip (get-font-size ctx *' 0.88) ++ read-inline ctx it ++ inline-fil)

let-inline ctx \dfn it =
  let ctx = ctx |> StdJaBook.set-cjk-font (`HiraKakuProN-W5`, 0.88, 0.) in
  read-inline ctx it

in


document (|
  title = {Test for CFF Subsets};
  author = {gfn};
  show-title = true;
  show-toc = false;
|) '<
    +p{
      これは既定のIPA明朝で組んだ段落です．TrueTypeアウトラインをもつOpenTypeフォントで，ずっと前から部分埋め込みが実現されています．
    }
    +p-hira{
      \href(`https://github.com/gfngfn/SATySFi`){\dfn{\SATySFi;}}は，2017年度IPA未踏事業のひとつとして開発された，新しい組版処理システムである。
      一口に組版処理システムといっても，大きく分けて

      \listing{
        * GUIの画面上に版面がプレヴューされた状態で文書を視覚的に記述していく\dfn{WYISIWYGエディタ}方式
        * \dfn{マークアップ言語}によってテキストファイルとして文書を記述し，
          言語処理系に入力として与えて最終的にPDFなどの形式で版面を出力する方式
      }%

      の2つがあるが，\SATySFi;は後者にあたる。すなわち，\LaTeX;やtroffなどと同様の形式をとるソフトウェアだ。
      では\LaTeX;をはじめとする既存のマークアップ言語方式と\SATySFi;との違いは何なのだろうか？
      簡潔に言ってしまえば，最大の違いは

      \listing{
        * コマンド定義を記述したコードの可読性が高くカスタマイズしやすい
        * ユーザが不適格な入力を与えたときのエラー報告がすばやく，わかりやすく提示される
      }%

      といった点にある。
      この性質の詳細を述べる前に，\SATySFi;を実装した動機をもう少し詳しく順を追って説明したい。
    }
>
```
where `HiraKakuProN-W(2|5)` are font names for ヒラギノ角ゴシック.

gfngfn commented:
I’m looking into the implementation of this PR and trying to find the cause of the problem (partly by refactoring the implementation like e26ea4dcb684b496cdebe9bbe57636329b5068e7 at branch `pullreq-matsud224-dev-cff-subset`), but still cannot tell what causes the problem.

matsud224 commented:
I reproduced the issue described above with Book App on iOS using NotoSansCJKjp-Regular instead of ヒラギノ角ゴシック.

matsud224 commented:
The problem may be solved by [this commit](https://github.com/matsud224/otfm/commit/ac7bec59a55b925b0ffa675d51a52901ed39f572). According to the CFF spec p.17, ROS or SyntheticBase operator in Top DICT should be put at the beginning of the DICT.

But I found another bug... Some PDF viewers (e.g. GNU gv) don't accept a zero-length data in INDEX. So, `remove_unused_subrs` should be fixed. I'm fixing this problem.

matsud224 commented:
I fixed the zero-length data bug (related discussion: https://bugs.chromium.org/p/chromium/issues/detail?id=69341) by [this commit](https://github.com/matsud224/otfm/commit/e9efb92401de5340a699afe1e4ac375be54a2f8f). To remove unused subroutines and reassign subroutine numbers are the best if possible. But it's hard to replace arguments of callsubr or callgsubr  (i.e. subroutine number) because it's not a immediate operand.

To fix this problem, I changed the dummy data length from zero to one.

gfngfn commented:
I am really sorry for the extremely late response… I confirmed that modifying the length of dummy data in subroutines from 0 to 1 works quite well! Here are rendered results of a PDF generated by the updated version of SATySFi which reflects the modification:

![スクリーンショット 2019-05-18 18 57 10](https://user-images.githubusercontent.com/6876129/57968039-58059580-79a0-11e9-97d4-13efe35e721a.png)

![スクリーンショット 2019-05-18 18 56 36](https://user-images.githubusercontent.com/6876129/57968043-6784de80-79a0-11e9-814a-7e6aa7e9eed0.png)

The first one was produced by `Preview.app`, and the second was by Adobe Acrobat Reader DC (for macOS).

I am deeply grateful to you for the improvement!
